File: ch5/src/py_demo_tf/py_demo_tf/static_tf_broadcaster.py
import math
import rclpy
from rclpy.node import Node
from tf2_ros import StaticTransformBroadcaster
from geometry_msgs.msg import TransformStamped
from tf_transformations import quaternion_from_euler

class StaticTFBrodcaster(Node):
    def __init__(self):
        super().__init__('static_tf2_broadcaster')
        self.static_broadcaster_ = StaticTransformBroadcaster(self)
        self.publish_static_tf()
        
    def publish_static_tf(self):
        transform = TransformStamped()
        transform.header.stamp = self.get_clock().now().to_msg()
        transform.header.frame_id = "base_link"
        transform.child_frame_id = "camera_link"
        transform.transform.translation.x = 0.5
        transform.transform.translation.y = 0.3
        transform.transform.translation.z = 0.6
        
        rotation_quaternion = quaternion_from_euler(math.radians(180), 0, 0)
        transform.transform.rotation.x = rotation_quaternion[0]
        transform.transform.rotation.y = rotation_quaternion[1]
        transform.transform.rotation.z = rotation_quaternion[2]
        transform.transform.rotation.w = rotation_quaternion[3]
        
        self.static_broadcaster_.sendTransform(transform)
        self.get_logger().info(f"publish TF : {transform}")
        
        
def main():
    rclpy.init()
    static_tf_broadcaster = StaticTFBrodcaster()
    rclpy.spin(static_tf_broadcaster)
    rclpy.shutdown()

File: ch5/src/cpp_demo_tf/src/static_tf_broadcaster.cpp
#include <memory>
#include "rclcpp/rclcpp.hpp"
#include "geometry_msgs/msg/transform_stamped.hpp"
#include "tf2/LinearMath/Quaternion.h"
#include "tf2_geometry_msgs/tf2_geometry_msgs.hpp"
#include "tf2_ros/static_transform_broadcaster.h"

class StaticTFBroadcaster: public rclcpp::Node {
private:
    std::shared_ptr<tf2_ros::StaticTransformBroadcaster> broadcaster_;

public:
    StaticTFBroadcaster(): Node("tf_broadcaster_node") {
        this->broadcaster_ = std::make_shared<tf2_ros::StaticTransformBroadcaster>(this);
        this->publish_tf();
    };

    void publish_tf() {
        geometry_msgs::msg::TransformStamped transform;
        transform.header.stamp = this->get_clock()->now();
        transform.header.frame_id = "map";
        transform.child_frame_id = "target_point";
        transform.transform.translation.x = 5.0;
        transform.transform.translation.y = 3.0;
        transform.transform.translation.z = 0.0;
        
        tf2::Quaternion quaternion;
        quaternion.setRPY(0, 0, 60 * M_PI /180);
        transform.transform.rotation = tf2::toMsg(quaternion);

        this->broadcaster_->sendTransform(transform);
    }
};

int main(int argc, char** argv) {
    rclcpp::init(argc, argv);
    auto tf_node = std::make_shared<StaticTFBroadcaster>();
    rclcpp::spin(tf_node);
    rclcpp::shutdown();

    return 0;
}